README
===========================================
+     _____        _    _                 +
+    |  __ \      (_)  | |                +
+    | |__) |_ __  _   | |____      __    +
+    |  _  /| '_ \| |  | '_ \ \ /\ / /    +
+    | | \ \| |_) | |==| | | \ V  V /     +
+    |_|  \_\ .__/|_|  |_| |_|\_/\_/      +
+           | |                           +
+           |_|                   v0.2.2  +
===========================================


Authors
=======
See AUTHORS file for more details.


License
=======
See COPYING file for more details.


ChangeLog
=========
See ChangeLog file for more details.


Library dependencies
====================
Rpi-hw has no dependencies.


Building with CMake
===================
 $ cd <SOURCES>
 $ cmake ./ -DCMAKE_INSTALL_PREFIX=<PATH>
 $ make
 # make install


Linking with Rpi-hw
=======================
To link:

 $ g++ `pkg-config --libs --cflags rpi-hw` <SOURCE> -o <BIN>


 


ChangeLog
ChangeLog of Rpi-hw - A free C++ library designed to manage the Raspberry Pi's GPIO connector and its other buses.

============ Rpi-hw 0.7.3 - urgency=high ===========

2014-03-02 Wicker25 < wicker25@gmail.com >

    * mcp23x08.hpp: Fixed addresses of the registers.

============ Rpi-hw 0.7.2 - urgency=low ===========

2013-12-26 Wicker25 < wicker25@gmail.com >

    * hd44780.hpp: Added support for `std::wstring`.

============ Rpi-hw 0.7.1 - urgency=low ===========

2013-12-24 Wicker25 < wicker25@gmail.com >

    * all: Added `__ENUM` and `__ENUM_PARAMS` macros.

    * utils.hpp: Added support for UTF-16 and UTF-32 string to the `utils::align`.

    * hd44780.hpp: Added support for Unicode strings.

============ Rpi-hw 0.7.0 - urgency=low ===========

2013-12-07 Wicker25 < wicker25@gmail.com >

    * all: Beautified all code! :)
           Added I2C support.
           Added `driver::mcp23017` to support 16-Bit I/O expander with I2C.
           Added `driver::mcp23008` and `driver::mcp23s08` to support 8-Bit I/O expander with I2C/SPI.
           Renamed `ioexpander.hpp` to `io-expander.hpp` and `ioexpander` to `io_expander`.
           Renamed `hc-sr04.hpp` to `hcsr04.hpp`.

    * iface/base.hpp: Renamed `numOfPins` to `size`.

    * display/hd44780.hpp: Improved performance and style.
                           Added `setTypingDelay` to set the typing delay between individual characters.
                           Added the text flags parameter to the function `write`.

============ Rpi-hw 0.6.0 - urgency=low ===========

2013-11-22 Wicker25 < wicker25@gmail.com >

    * all: Added directory `driver` containing the main hardware drivers.
           Renamed `gpio` to `driver::bcm2835`.
           Added `driver::mcp23s17` to support 16-Bit I/O expander with SPI.
           Added new class `gpio` to handles all GPIOs.
           Removed class `pcd8544-skeleton` and `pcd8544-spi`.
           Removed `rpi-gpio`.
           Improved names of enumeration constants.
           Renamed `PUD_OFF` to `PULL_OFF`.
           Renamed `pud_mode` to `pull_mode`.

    * keypad/base.hpp: Changed the input polarity using pull-up resistors (now it works with I/O expanders).

============ Rpi-hw 0.5.2 - urgency=medium ===========

2013-10-22 Wicker25 < wicker25@gmail.com >

    * keypad/base.hpp: Added the refresh rate parameter and support for event listener.

============ Rpi-hw 0.5.1 - urgency=medium ===========

2013-09-20 Wicker25 < wicker25@gmail.com >

    * all: Moved SPI sources into the directory `rpi-hw/iface`.
           Added SPI support to the PCD8544 controller.

============ Rpi-hw 0.5.0 - urgency=low =============

2013-09-15 Wicker25 < wicker25@gmail.com >

    * all: Added SPI support.

============ Rpi-hw 0.4.2 - urgency=high =============

2013-08-07 Wicker25 < wicker25@gmail.com >

    * keypad/base.cpp: Fixed bad initialization.
                       Added keymap support.

============ Rpi-hw 0.4.1 - urgency=low =============

2013-05-06 Wicker25 < wicker25@gmail.com >

    * all: Added class `hcsr04` to manage HC-SR04 ultrasonic distance sensors.

    * gpio.hpp: Replaced `m_addr` with `m_gpio`.

    * utils.hpp: Rewrote the function `nsleep`.

============ Rpi-hw 0.4.0 - urgency=low =============

2013-05-06 Wicker25 < wicker25@gmail.com >

    * all: Welcome C++11!
           Replaced UTF8-CPP with C++ built-in UTF-32.
           Replaced `bitset` with `std::vector< bool >` specialization.

    * bitmap.hpp, designer.hpp: Replaced the type of the bitmap buffer.

    * freetype.cpp: Added file checking.

    * hd44780.cpp: Fixed line addressing.

============ Rpi-hw 0.3.1 - urgency=medium =============

2013-05-01 Wicker25 < wicker25@gmail.com >

    * all: Fixed `rpi-hw.pc` and splitted CMake's files.
           Now headers of inline functions are automatically included.

============ Rpi-hw 0.3 - urgency=low =============

2013-04-28 Wicker25 < wicker25@gmail.com >

    * all: Added class `pcd8544` to manage PCD8544 controller.
           Added class `stepper` to manage stepper motors.
           Added class `designer`to draw shapes, images, texts, etc.
           Added Unicode support (using UTF8-CPP).
           Added Images support (using GraphicsMagick++).
           Added Fonts support (using FreeType).
           Moved methods `sleep`, `msleep`, and `usleep` into the namespace `time`.
           Improved virtual interface classes.

    * utils.hpp: Renamed `mcopy` to `memdup`.
                 Improved word wrap algorithm.

    * CMakeLists.txt: Improved! ;D

============ Rpi-hw 0.2.5 - urgency=high =============

2012-12-17 Wicker25 < wicker25@gmail.com >

    * CMakeLists.txt: Fixed recognition of the RPi Revision.

    * rpi-hw.pc.in: Fixed the path to the header files.

============ Rpi-hw 0.2.4 - urgency=low =============

2012-11-28 Wicker25 < wicker25@gmail.com >

    * doxy.conf.in: Updated to the Doxygen's version 1.8.1.2.

============ Rpi-hw 0.2.3 - urgency=low =============

2012-11-27 Wicker25 < wicker25@gmail.com >

    * all: Added Doxygen documentation (use `make doc` to build it!).
           Changed `README` to `README.md`.

============ Rpi-hw 0.2.2 - urgency=low =============

2012-11-18 Wicker25 < wicker25@gmail.com >

    * all: Added class `hd44780` to manage LCD displays.
           Renamed `cpu.hpp` to `rpi.hpp`.
           Renamed `macro.hpp` to `utils.hpp`.
           Improved base interface classes.

    * iface/base.hpp: Improved methods to manage GPIOs.

    * iface/utils.hpp: Added methods `malloc` and `memset`.
                       Added methods `sleep`, `msleep`, and `usleep`.
                       Added method `align` to align a text.

============ Rpi-hw 0.2.1 - urgency=low =============

2012-10-26 Wicker25 < wicker25@gmail.com >

    * all: Translated all comments to english - I hope they are clear!
           Improved error messages.

============ Rpi-hw 0.2.0 - urgency=low =============

2012-10-24 Wicker25 < wicker25@gmail.com >

src/rpi-gpio.cpp
/* 
    Title --- rpi-gpio.cpp [tools]

    Copyright (C) 2013 Giacomo Trudu - wicker25[at]gmail[dot]com

    This file is part of Rpi-hw.

    Rpi-hw is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation version 3 of the License.

    Rpi-hw is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with Rpi-hw. If not, see <http://www.gnu.org/licenses/>.
*/


#include <iostream>
#include <cstdlib>
#include <cstring>

#include <rpi-hw.hpp>
#include <rpi-hw/iface/gpio.hpp>

// Constants
#define RPI_GPIO_VERSION	"0.2"

static const char *usage =
"Usage: rpi-gpio <set|write|event|write|read> <gpio> [options]\n\n"
"       rpi-gpio set   <gpio> <in|out|pulldown|pullup|pudoff>\n"
"       rpi-gpio write <gpio> <high|low>\n"
"       rpi-gpio event <gpio> <rising|falling|high|low> <enabled|disabled>\n"
"       rpi-gpio event <gpio>\n"
"       rpi-gpio read  <gpio>\n"
"       rpi-gpio -h\n"
"       rpi-gpio -v\n";


// Raspberry Pi pins
#if RPI_REVISION == 0x0002 || RPI_REVISION == 0x0003

//                       pin  =    0   1   2   3   4   5   6   7   8   9  10  11  12  13
static const int pin_gpio[27] = { -1, -1, -1,  0, -1,  1, -1,  4, 14, -1, 15, 17, 18, 21,
//								  14  15  16  17  18  19  20  21 22  23  24  25  26
								  -1, 22, 23, -1, 24, 10, -1, 9, 25, 11,  8, -1,  7 };

#else

//                       pin  =    0   1   2  3   4   5   6   7   8   9  10  11  12  13
static const int pin_gpio[27] = { -1, -1, -1, 2, -1,  3, -1,  4, 14, -1, 15, 17, 18, 27,
//								  14  15  16  17  18  19  20  21  22  23  24  25  26
								  -1, 22, 23, -1, 24, 10, -1,  9, 25, 11,  8, -1,  7 };
#endif


// Use the Rpi-hw namespace
using namespace rpihw;

int
main( int argc, char *args[] ) {

	// Control flag
	bool success = false;

	// Check minimum number of arguments
	if ( argc > 2 ) {

		// GPIO control interface
		iface::gpio io;

		// Get selected GPIO pin
	    char *garbage = NULL;
		int pin;

		if ( !strncasecmp( args[2], "pin", 3 ) ) {

			pin = (int) strtol( ( args[2] + 3 ), &garbage, 0 );
			pin = ( *garbage == '\0' && pin >= 0 && pin < 27 ) ? pin_gpio[pin] : -1;

		} else {

			pin = (int) strtol( args[2], &garbage, 0 );
			if ( *garbage != '\0' || pin >= NUM_OF_PINS ) pin = -1;
		}

		// Check if there are no errors
		if ( pin != -1 ) {

			// Set GPIO pin mode
			if ( !strcasecmp( args[1], "set" ) ) {

				if ( argc > 3 ) {

					// Input mode
					if ( !strcasecmp( args[3], "in" ) || !strcasecmp( args[3], "input" ) ) {

						io.setup( (uint8_t) pin, iface::gpio::INPUT );
						success = true;

					// Output mode
					} else if ( !strcasecmp( args[3], "out" ) || !strcasecmp( args[3], "output" ) ) {

						io.setup( (uint8_t) pin, iface::gpio::OUTPUT );
						success = true;

					// Pull-down control
					} else if ( !strcasecmp( args[3], "pulldown" ) ) {

						io.setPullUpDown( (uint8_t) pin, iface::gpio::PULL_DOWN );
						success = true;

					// Pull-up control
					} else if ( !strcasecmp( args[3], "pullup" ) ) {

						io.setPullUpDown( (uint8_t) pin, iface::gpio::PULL_UP );
						success = true;

					// Disable pull-up/down control
					} else if ( !strcasecmp( args[3], "pudoff" ) ) {

						io.setPullUpDown( (uint8_t) pin, iface::gpio::PUD_OFF );
						success = true;
					}
				}

			// Set events
			} else if ( !strcasecmp( args[1], "event" ) ) {

				if ( argc > 3 ) {

					if ( argc > 4 ) {

						// Get selected mode
						int mode = -1;

						if ( !strcasecmp( args[4], "enabled" ) )
							mode = 1;
						else if ( !strcasecmp( args[4], "disabled" ) )
							mode = 0;

						// Check if there are no errors
						if ( mode != -1 ) {

							// Rising edge event
							if ( !strcasecmp( args[3], "rising" ) ) {

								io.setRisingEvent( (uint8_t) pin, (bool) mode );
								success = true;

							// Falling edge event
							} else if ( !strcasecmp( args[3], "falling" ) ) {

								io.setFallingEvent( (uint8_t) pin, (bool) mode );
								success = true;

							// High event
							} else if ( !strcasecmp( args[3], "high" ) ) {

								io.setHighEvent( (uint8_t) pin, (bool) mode );
								success = true;

							// Low event
							} else if ( !strcasecmp( args[3], "low" ) ) {

								io.setLowEvent( (uint8_t) pin, (bool) mode );
								success = true;
							}
						}
					}

				// Otherwise, detect events 
				} else {

					std::cout << io.checkEvent( (uint8_t) pin ) << std::endl;
					success = true;
				}

			// Write on output pin
			} else if ( !strcasecmp( args[1], "write" ) ) {

				// Get the value of the output pin
				int value = -1;

				if ( argc > 3 ) {

					if ( *args[3] == '0' || !strcasecmp( args[3], "low" ) )
						value = iface::gpio::LOW;
					else if ( *args[3] == '1' || !strcasecmp( args[3], "high" ) )
						value = iface::gpio::HIGH;
				}

				// Check if there are errors
				if ( value != -1 ) {

					// Write the value on the GPIO controller interface
					io.write( (uint8_t) pin, value );
					success = true;
				}

			// Read from input pin
			} else if ( !strcasecmp( args[1], "read" ) ) {

				// Read the value from the GPIO controller interface
				std::cout << io.read( (uint8_t) pin ) << std::endl;
				success = true;
			}
		}
	}

	if ( argc > 1 ) {

		// Print `rpi-gpio` version
		if ( !strcasecmp( args[1], "-v" ) ) {

			std::cout << "rpi-gpio v" << RPI_GPIO_VERSION << " by Giacomo Trudu aka `Wicker25`" << std::endl;
			success = true;

		// Print helps
		} else if ( !strcasecmp( args[1], "-h" ) ) {

			std::cout << usage << std::endl;
			success = true;
		}
	}

	// Handle other error
	if ( !success ) {

		std::cout << usage << std::endl;
	}

	return 0;
}
